File: apps/desktop/src/lib/feed.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * The Home feed and profiles, as the WebView sees them.
 *
 * **None of this is end-to-end encrypted**, and that is not a caveat buried in
 * a type file — §4.4 requires the product to say it in plain language where
 * someone is about to post, which `HomePage` does under the composer. It is
 * repeated here because this is the module a future change would touch while
 * assuming the rest of the app's guarantees carry over. They do not: the server
 * can read every post, every profile field, and every feed image.
 *
 * What still holds is rule 2. Image bytes and presigned URLs stay in Rust: the
 * WebView passes a **path** to `uploadImage` and gets back an object key.
 */

export interface Post {
  id: number;
  author_id: number;
  author_handle: string;
  author_display_name: string;
  author_avatar_key: string | null;
  body: string;
  /** Object keys, not URLs. Each is presigned on demand by `imageUrl`. */
  media_keys: string[];
  created_at_ms: number;
  reactions: ReactionCount[];
  /** Emoji this account has used, so the UI can draw them pressed. */
  my_reactions: string[];
  is_mine: boolean;
  /** The headline, when there is one. Posts written before titles have none. */
  title: string | null;
  kind: PostKind;
  /** Where a link post points. `null` for the other kinds. */
  link_url: string | null;
  /** Upvotes minus downvotes. */
  score: number;
  /** This account's own vote: 1, -1, or 0. */
  my_vote: number;
  comment_count: number;
  /** Pinned to the top of its author's profile. Only set on a profile page. */
  pinned?: boolean;
}

export type PostKind = "text" | "link" | "image";

/** How a feed page is ordered. */
export type FeedSort = "new" | "top" | "hot";

/**
 * One comment, flat.
 *
 * The tree is rebuilt from `parent_id` where it is drawn, rather than arriving
 * nested: a flat list keeps the response shape independent of depth, and the
 * whole thread comes in one round trip either way.
 */
export interface Comment {
  id: number;
  post_id: number;
  /** `null` at the top level. */
  parent_id: number | null;
  author_id: number;
  author_handle: string;
  author_display_name: string;
  author_avatar_key: string | null;
  /** Empty when deleted. */
  body: string;
  created_at_ms: number;
  is_mine: boolean;
  /** Deleted comments keep their place so their replies keep theirs. */
  deleted: boolean;
}

export interface VoteResult {
  score: number;
  my_vote: number;
}

export interface ReactionCount {
  emoji: string;
  count: number;
}

export interface FeedPage {
  posts: Post[];
  /** Pass back as `before`. `null` means the end of the feed. */
  next_cursor: number | null;
}

export interface ProfileLink {
  label: string;
  /** Always http(s) — refused at three layers, and opened in the system
   * browser rather than the WebView. */
  url: string;
}

/**
 * Someone's profile, as far as this viewer may see it.
 *
 * A field that is `null` is **hidden**, not empty. The UI must say "not
 * shared" rather than drawing a blank line that reads as "they wrote nothing" —
 * the two are different facts and conflating them misrepresents a privacy
 * setting as an empty one.
 */
export interface Profile {
  user_id: number;
  handle: string;
  display_name: string;
  avatar_key: string | null;
  banner_key: string | null;
  bio: string | null;
  location: string | null;
  links: ProfileLink[] | null;
  join_date_ms: number | null;
  is_me: boolean;
}

/** Who may see one field (G2). */
export type Visibility = "public" | "contacts" | "private";

/** A settable field. Handle and display name are absent: they are how you are
 * addressed, so a control for them would be one that cannot be honoured. */
export type VisibilityField = "bio" | "location" | "links" | "join_date";

export interface MyProfile extends Profile {
  /** Every settable field, with the value actually in force. */
  visibility: Record<VisibilityField, Visibility>;
}

export interface ProfileEdit {
  display_name?: string;
  bio?: string;
  location?: string;
  links?: ProfileLink[];
  avatar_key?: string;
  banner_key?: string;
}

export interface FeedError {
  kind:
    | "unreachable"
    | "signed_out"
    | "rejected"
    | "invalid_request"
    | "unreadable_file"
    | "too_large"
    | "internal";
  message: string;
}

/** Narrows an unknown rejection to something renderable. */
export function asFeedError(error: unknown): FeedError {
  if (
    typeof error === "object" &&
    error !== null &&
    "kind" in error &&
    "message" in error
  ) {
    return error as FeedError;
  }
  return { kind: "internal", message: "Something went wrong. Try again." };
}

/**
 * A page of the feed.
 *
 * `before` is a post id under `new` and a row offset under `top` and `hot` —
 * those orders are not monotonic in id, so an id cannot express a page
 * boundary in them. Pass back whatever `next_cursor` gave you.
 */
export function feed(before?: number, sort: FeedSort = "new"): Promise<FeedPage> {
  return invoke<FeedPage>("feed", { before: before ?? null, sort });
}

export function postsBy(handle: string, before?: number): Promise<FeedPage> {
  return invoke<FeedPage>("posts_by", { handle, before: before ?? null });
}

export function createPost(input: {
  body: string;
  mediaKeys?: string[];
  title?: string | null;
  kind?: PostKind;
  linkUrl?: string | null;
}): Promise<Post> {
  return invoke<Post>("create_post", {
    body: input.body,
    mediaKeys: input.mediaKeys ?? [],
    title: input.title ?? null,
    kind: input.kind ?? "text",
    linkUrl: input.linkUrl ?? null,
  });
}

/** Pins one of your own posts. At most three; the server enforces it. */
export function pinPost(id: number): Promise<void> {
  return invoke<void>("pin_post", { id });
}

/** Unpins one of your own posts. */
export function unpinPost(id: number): Promise<void> {
  return invoke<void>("unpin_post", { id });
}

export function deletePost(id: number): Promise<void> {
  return invoke<void>("delete_post", { id });
}

export function react(
  id: number,
  emoji: string,
  on: boolean,
): Promise<ReactionCount[]> {
  return invoke<ReactionCount[]>("react", { id, emoji, on });
}

/** Votes on a post. `1`, `-1`, or `0` to take a vote back. */
export function vote(id: number, value: number): Promise<VoteResult> {
  return invoke<VoteResult>("vote", { id, value });
}

/** The whole thread for a post, oldest first, deleted ones included. */
export function comments(postId: number): Promise<Comment[]> {
  return invoke<Comment[]>("comments", { postId });
}

/** Adds a comment, or a reply when `parentId` is given. */
export function addComment(
  postId: number,
  body: string,
  parentId?: number | null,
): Promise<Comment> {
  return invoke<Comment>("add_comment", { postId, body, parentId: parentId ?? null });
}

export function deleteComment(id: number): Promise<void> {
  return invoke<void>("delete_comment", { id });
}

export function profile(handle: string): Promise<Profile> {
  return invoke<Profile>("profile", { handle });
}

export function myProfile(): Promise<MyProfile> {
  return invoke<MyProfile>("my_profile");
}

export function updateProfile(edit: ProfileEdit): Promise<MyProfile> {
  return invoke<MyProfile>("update_profile", { edit });
}

export function updateVisibility(
  visibility: Partial<Record<VisibilityField, Visibility>>,
): Promise<MyProfile> {
  return invoke<MyProfile>("update_visibility", { visibility });
}

/**
 * Uploads an image the user already picked, and returns its object key.
 *
 * The path goes to Rust; the bytes and the presigned URL stay there. A
 * presigned PUT is a bearer credential for one object, and there is no reason
 * for a WebView to be holding one.
 */
export function uploadImage(path: string): Promise<string> {
  return invoke<string>("upload_image", { path });
}

/**
 * A stored image, inlined as a `data:` URL.
 *
 * Not the presigned object-storage URL it looks like it should be: the CSP
 * allows `img-src 'self' asset: data: blob:` and no remote host, so a bucket
 * URL is blocked before a byte is fetched. Rust downloads it and hands over
 * the bytes, which keeps the bucket unreachable from anything running in the
 * page.
 */
export function imageUrl(key: string): Promise<string> {
  return invoke<string>("image_data_url", { key });
}

File: crates/client/src/feed.rs
//! The Home feed and profiles, client side (§6.2, §6.3).
//!
//! Deliberately separate from [`crate::transport::Transport`]. That trait is
//! the end-to-end encrypted path — key packages, envelopes, commits — and it is
//! a trait so the parts that matter can be tested against a fake without a
//! server. None of that applies here: a feed post is plaintext both ends, there
//! is no MLS state to keep, and nothing to get wrong that a fake would catch.
//! Bolting ten more methods onto `Transport` would make every implementation
//! stub out ten things it does not care about, and blur the one distinction the
//! whole codebase is organised around.
//!
//! **Nothing in this module is end-to-end encrypted.** §4.4 requires that to be
//! said in the UI, in plain language, where someone is about to post — see
//! `FeedNotice`. Saying it here too, because this is the file where somebody
//! would otherwise assume the rest of the app's guarantees carry over.

use serde::{Deserialize, Serialize};

use crate::transport::TransportError;

/// One post as the feed renders it.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Post {
    /// Server-assigned. Also the paging cursor.
    pub id: i64,
    /// Who wrote it.
    pub author_id: i64,
    /// Their handle, for `@name` and for opening their profile.
    pub author_handle: String,
    /// Their display name.
    pub author_display_name: String,
    /// Their avatar's object key, if they have one.
    pub author_avatar_key: Option<String>,
    /// The text. Server-readable, like everything else here.
    pub body: String,
    /// Object keys in `nexo-media`, presigned individually to render.
    pub media_keys: Vec<String>,
    /// When the server accepted it, in milliseconds since the epoch.
    pub created_at_ms: i64,
    /// Every emoji used on it, with counts.
    pub reactions: Vec<ReactionCount>,
    /// Emoji this account has used, so the UI can show them pressed.
    pub my_reactions: Vec<String>,
    /// True when this account wrote it — the only one who may delete it.
    pub is_mine: bool,
    /// The headline, when there is one. Posts written before titles have none.
    #[serde(default)]
    pub title: Option<String>,
    /// `text`, `link` or `image`.
    #[serde(default = "text_kind")]
    pub kind: String,
    /// Where a link post points. `None` for the other kinds.
    #[serde(default)]
    pub link_url: Option<String>,
    /// Upvotes minus downvotes.
    #[serde(default)]
    pub score: i64,
    /// This account's own vote: `1`, `-1`, or `0`.
    #[serde(default)]
    pub my_vote: i16,
    /// How many comments the thread holds.
    #[serde(default)]
    pub comment_count: i64,
    /// Pinned to the top of its author's profile. Only ever set on a
    /// profile's own page; the global feed does not treat pinning as a signal.
    #[serde(default)]
    pub pinned: bool,
}

/// What a post is when a server predates post kinds.
fn text_kind() -> String {
    "text".to_string()
}

/// One comment, flat. The tree is rebuilt from `parent_id` by whoever draws it.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Comment {
    /// Server-assigned.
    pub id: i64,
    /// Which post it is on.
    pub post_id: i64,
    /// The comment being replied to. `None` at the top level.
    pub parent_id: Option<i64>,
    /// Who wrote it.
    pub author_id: i64,
    /// Their handle.
    pub author_handle: String,
    /// Their display name.
    pub author_display_name: String,
    /// Their avatar's object key, if any.
    pub author_avatar_key: Option<String>,
    /// Empty when deleted.
    pub body: String,
    /// When the server accepted it, in milliseconds since the epoch.
    pub created_at_ms: i64,
    /// True when this account wrote it.
    pub is_mine: bool,
    /// Deleted comments keep their place so their replies keep theirs.
    pub deleted: bool,
}

/// What a post is being written as.
#[derive(Debug, Clone, Default, Serialize)]
pub struct NewPost {
    /// The headline. Optional.
    pub title: Option<String>,
    /// `text`, `link` or `image`.
    pub kind: String,
    /// Required for a link post.
    pub link_url: Option<String>,
    /// The text.
    pub body: String,
    /// Object keys already uploaded.
    pub media_keys: Vec<String>,
}

/// The score after a vote.
#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
pub struct VoteResult {
    /// Upvotes minus downvotes.
    pub score: i64,
    /// What this account's vote now is.
    pub my_vote: i16,
}

/// One emoji and how many people used it.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ReactionCount {
    /// The emoji itself.
    pub emoji: String,
    /// How many people used it.
    pub count: i64,
}

/// A page of the feed.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct FeedPage {
    /// Newest first.
    pub posts: Vec<Post>,
    /// Pass back as `before`. `None` means the end.
    pub next_cursor: Option<i64>,
}

/// A link on a profile.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ProfileLink {
    /// What to show.
    pub label: String,
    /// Always `http://` or `https://` — the server refuses anything else, and
    /// the UI opens it in the system browser rather than the WebView.
    pub url: String,
}

/// A profile, as far as this viewer is allowed to see it.
///
/// A field the viewer may not see is `None`, not an empty string. The
/// difference is what lets the UI say "not shared" instead of drawing a blank
/// line that reads as "they wrote nothing".
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Profile {
    /// The in-app numeric id (§3: never a phone number).
    pub user_id: i64,
    /// The unique handle.
    pub handle: String,
    /// What to call them.
    pub display_name: String,
    /// Object key for the avatar.
    pub avatar_key: Option<String>,
    /// Object key for the 3:1 banner.
    pub banner_key: Option<String>,
    /// `None` when hidden from this viewer, not when empty.
    pub bio: Option<String>,
    /// Free text. `None` when hidden — which is the default.
    pub location: Option<String>,
    /// `None` when hidden.
    pub links: Option<Vec<ProfileLink>>,
    /// Milliseconds since the epoch. `None` when hidden.
    pub join_date_ms: Option<i64>,
    /// True when this is the signed-in account's own profile.
    pub is_me: bool,
}

/// Your own profile, plus the visibility settings only you can see.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct MyProfile {
    /// The profile itself, with nothing hidden.
    #[serde(flatten)]
    pub profile: Profile,
    /// Every settable field, with the value actually in force.
    pub visibility: std::collections::BTreeMap<String, String>,
}

/// What to change about your own profile. Absent fields are left alone.
#[derive(Debug, Clone, Default, Serialize)]
pub struct ProfileEdit {
    /// 1 to 40 characters.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    /// Up to 280 characters (§6.3).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bio: Option<String>,
    /// Free text. Never a geolocation API (§6.3).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub location: Option<String>,
    /// Replaces the whole list when present.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub links: Option<Vec<ProfileLink>>,
    /// A key already uploaded to `nexo-media`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub avatar_key: Option<String>,
    /// Same, for the banner.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub banner_key: Option<String>,
}

/// One account you are blocking, as the settings list shows it.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Block {
    /// Their handle, which is how you unblock them again.
    pub handle: String,
    /// What they call themselves, so the list is readable.
    pub display_name: String,
    /// When the block was made. The list is newest first.
    pub blocked_at_ms: i64,
}

/// The feed and profile API.
///
/// A trait for the same reason `Transport` is one — so a caller depends on the
/// shape rather than on `ureq` — but with none of the crypto, and no fake in
/// the tree because there is nothing here a fake would prove.
pub trait FeedApi {
    /// A page of the global feed.
    ///
    /// `sort` is `new` (default), `top` or `hot`. `before` is a post id under
    /// `new` and a row offset under the others — those orders are not monotonic
    /// in id, so an id cannot express a page boundary in them. Pass back
    /// whatever `next_cursor` returned and it is right either way.
    fn feed(
        &self,
        before: Option<i64>,
        limit: Option<i64>,
        sort: Option<&str>,
    ) -> Result<FeedPage, TransportError>;

    /// One person's posts.
    fn posts_by(
        &self,
        handle: &str,
        before: Option<i64>,
        limit: Option<i64>,
    ) -> Result<FeedPage, TransportError>;

    /// Writes a post.
    fn create_post(&self, new: &NewPost) -> Result<Post, TransportError>;

    /// Deletes one of your own.
    fn delete_post(&self, id: i64) -> Result<(), TransportError>;

    /// Adds or removes one reaction. Returns the new counts.
    fn react(&self, id: i64, emoji: &str, on: bool) -> Result<Vec<ReactionCount>, TransportError>;

    /// Votes on a post. `1`, `-1`, or `0` to take a vote back.
    fn vote(&self, id: i64, value: i16) -> Result<VoteResult, TransportError>;

    /// The whole comment thread for a post, oldest first.
    fn comments(&self, post_id: i64) -> Result<Vec<Comment>, TransportError>;

    /// Adds a comment, or a reply when `parent_id` is set.
    fn add_comment(
        &self,
        post_id: i64,
        body: &str,
        parent_id: Option<i64>,
    ) -> Result<Comment, TransportError>;

    /// Deletes one of your own comments.
    fn delete_comment(&self, id: i64) -> Result<(), TransportError>;

    /// Somebody's public profile.
    fn profile(&self, handle: &str) -> Result<Profile, TransportError>;

    /// Your own, with nothing hidden.
    fn my_profile(&self) -> Result<MyProfile, TransportError>;

    /// Edits your own.
    fn update_profile(&self, edit: &ProfileEdit) -> Result<MyProfile, TransportError>;

    /// Sets who may see which fields (G2).
    fn update_visibility(
        &self,
        visibility: &std::collections::BTreeMap<String, String>,
    ) -> Result<MyProfile, TransportError>;

    /// Pins one of your own posts to the top of your profile.
    fn pin_post(&self, id: i64) -> Result<(), TransportError>;

    /// Unpins one of your own. Idempotent.
    fn unpin_post(&self, id: i64) -> Result<(), TransportError>;

    /// Everyone you are blocking.
    ///
    /// Only your own list. There is deliberately no way to ask who is blocking
    /// *you* -- see `blocks.rs` on the server for why.
    fn blocks(&self) -> Result<Vec<Block>, TransportError>;

    /// Blocks somebody. Idempotent.
    fn block(&self, handle: &str) -> Result<(), TransportError>;

    /// Unblocks somebody. Idempotent.
    fn unblock(&self, handle: &str) -> Result<(), TransportError>;

    /// A time-limited URL to PUT a feed or profile image to, and its key.
    fn media_upload_url(&self, size: u64) -> Result<(String, String), TransportError>;

    /// A time-limited URL to GET one from.
    fn media_download_url(&self, key: &str) -> Result<String, TransportError>;
}
